From 07ee90b8275dda14e06db4f0640213d58918d008 Mon Sep 17 00:00:00 2001
From: cheng_cai <cheng_cai@realsil.com.cn>
Date: Fri, 15 Jun 2018 14:50:28 +0800
Subject: [PATCH] BT Interface: fix hci cmd timeout

Add mutex in function Send() and OnTimeout() to do mutual
exclusive access on lpm_wake_deasserted.

If lpm_wake_deasserted is out of sync with the actual
wake state, then platforms which use the lpm hint
will break.

Bug: 64299848
Test: run affected platform for a long period of time
and find that hci commands do not time out

Change-Id: I9b83cdc5251f78e0ceb0aea8e7887ee2da441c0e
---
 bluetooth/1.0/default/vendor_interface.cc | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/bluetooth/1.0/default/vendor_interface.cc b/bluetooth/1.0/default/vendor_interface.cc
index 44ca5608..1af8bf7a 100755
--- a/bluetooth/1.0/default/vendor_interface.cc
+++ b/bluetooth/1.0/default/vendor_interface.cc
@@ -51,6 +51,7 @@ uint32_t lpm_timeout_ms;
 bool recent_activity_flag;
 
 VendorInterface* g_vendor_interface = nullptr;
+std::mutex wakeup_mutex_;
 
 HC_BT_HDR* WrapPacketAndCopy(uint16_t event, const hidl_vec<uint8_t>& data) {
   size_t packet_size = data.size() + sizeof(HC_BT_HDR);
@@ -310,6 +311,7 @@ void VendorInterface::Close() {
 }
 
 size_t VendorInterface::Send(uint8_t type, const uint8_t* data, size_t length) {
+  std::unique_lock<std::mutex> lock(wakeup_mutex_);
   recent_activity_flag = true;
 
   if (lpm_wake_deasserted == true) {
@@ -352,6 +354,7 @@ void VendorInterface::OnFirmwareConfigured(uint8_t result) {
 
 void VendorInterface::OnTimeout() {
   ALOGV("%s", __func__);
+  std::unique_lock<std::mutex> lock(wakeup_mutex_);
   if (recent_activity_flag == false) {
     lpm_wake_deasserted = true;
     bt_vendor_lpm_wake_state_t wakeState = BT_VND_LPM_WAKE_DEASSERT;
-- 
2.25.1

